### PR TITLE
Fix metrics drop when using CoreOS kernel >= 1122.2.0

### PR DESCRIPTION
**Pull request by richardpen**
### Summary

Fix an issue where agent get wrong cpu number for metrics when running on CoreOS with kernel version >= 1122.2.0
### Implementation details

Using runtime.NumCPU() to get the number of cpu cores instead of depending on entries of percpu_usage returned by docker stats
### Testing

- [x] Builds (`make release`)
- [x] Unit tests (`make test`) pass
- [x] Integration tests (`make run-integ-tests`) pass
- [x] Functional tests (`make run-functional-tests`) pass
- [x] Manually test on CoreOS with both the kernel 1122.2.0 and 1068.9.0

New tests cover the changes: 
yes
### Description for the changelog

no change log
### Licensing

This contribution is under the terms of the Apache 2.0 License: 
yes

### CODE REVIEW

File: agent/stats/utils.go [hunk 1]
```diff
@@ -35,9 +38,8 @@ func nan32() float32 {
 // dockerStatsToContainerStats returns a new object of the ContainerStats object from docker stats.
 func dockerStatsToContainerStats(dockerStats *docker.Stats) (*ContainerStats, error) {
 	// The length of PercpuUsage represents the number of cores in an instance.
-	numCores := uint64(len(dockerStats.CPUStats.CPUUsage.PercpuUsage))
-	if numCores == 0 {
-		seelog.Debug("Invalid container statitistics reported, got number of cores = 0")
+	if len(dockerStats.CPUStats.CPUUsage.PercpuUsage) == 0 {
+		seelog.Debug("Invalid container statitistics reported, got nil for cpu usage")
```
Comment (denderello) [inline agent/stats/utils.go]: Looks like you have got a typo here. :) `statitistics ` => `statistics`

Comment (richardpen) [inline agent/stats/utils.go]: Thanks for catching, fixed

File: agent/stats/utils.go [hunk 2]
```diff
@@ -35,9 +38,8 @@ func nan32() float32 {
 // dockerStatsToContainerStats returns a new object of the ContainerStats object from docker stats.
 func dockerStatsToContainerStats(dockerStats *docker.Stats) (*ContainerStats, error) {
 	// The length of PercpuUsage represents the number of cores in an instance.
-	numCores := uint64(len(dockerStats.CPUStats.CPUUsage.PercpuUsage))
-	if numCores == 0 {
-		seelog.Debug("Invalid container statitistics reported, got number of cores = 0")
+	if len(dockerStats.CPUStats.CPUUsage.PercpuUsage) == 0 {
```
Comment (aaithal) [inline agent/stats/utils.go]: Do we still need this check here? When `dockerStats.CPUStats.CPUUsage.PercpuUsage` is `0`, do rest of the fields in the struct have invalid values as well? Would be good to run a `stress` workload and inspect the output when this happens.

File: agent/stats/utils_test.go [hunk 3]
```diff
@@ -38,6 +38,7 @@ func TestIsNetworkStatsError(t *testing.T) {
 func TestDockerStatsToContainerStatsCpuUsage(t *testing.T) {
 	// doing this with json makes me sad, but is the easiest way to deal with
 	// the inner structs
+	numCores = 4
```
Comment (samuelkarp) [inline agent/stats/utils_test.go]: Can you add a note that this is changing a package-level variable?  It's hiding here like a locally-scoped variable and is non-obvious that it matters.

File: agent/stats/utils.go (hunk 2 above)
Comment (richardpen) [inline agent/stats/utils.go]: From the stress [output](https://gist.github.com/richardpen/cb80f285ab558e79a4a8ea0a116cf629), if `len(dockerStats.CPUStats.CPUUsage.PercpuUsage)` is `0`, the data returned is useless, so it should be ignored.

File: agent/stats/utils.go [hunk 4]
```diff
@@ -35,9 +38,8 @@ func nan32() float32 {
...
+		seelog.Debug("Invalid container statistics reported, got nil for cpu usage")
```
Comment (aaithal) [inline agent/stats/utils.go]: Can you modify the message to not contain `nil`? Something like `Invalid container statistics reported, invalid stats payload from docker`?

**Comment by richardpen** [comment]
@aaithal Sam's comment has been addressed in the [code](https://github.com/aws/amazon-ecs-agent/pull/556/files#diff-13afca01e3a8339ae7856f996a5b1859R41).

**Comment by richardpen** [comment]
Merged

**Comment by mlafeldt** [comment]
Thanks for the fix, @richardpen. Now looking forward to a new ecs-agent release. 😄

**Comment by mlafeldt** [comment]
Hey, would it be possible to release this fix as version 1.13.1? Or do you have other plans, e.g. cutting 1.14 with more changes?

**Comment by samuelkarp** [comment]
@richardpen One note on this: we should have a changelog entry addressing this as a bug fix.

Something like:

Bug - Resolved a bug where CPU metrics would be calculated incorrectly for some kernels.

**Comment by richardpen** [comment]
@samuelkarp thanks, updated in the summary.

**Pull request closed without merging**